daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala

```diff
@@ -436,16 +436,16 @@ sealed abstract class HasTxNodes {
 
   /** Returns the IDs of all input contracts that are used by this transaction.
     */
-  final def inputContracts[Cid2 >: ContractId]: Set[Cid2] =
-    fold(Set.empty[Cid2]) {
-      case (acc, (_, Node.Exercise(coid, _, _, _, _, _, _, _, _, _, _, _, _, _, _))) =>
-        acc + coid
-      case (acc, (_, Node.Fetch(coid, _, _, _, _, _, _, _))) =>
-        acc + coid
-      case (acc, (_, Node.LookupByKey(_, _, Some(coid), _))) =>
-        acc + coid
+  final def inputContracts[Cid2 >: ContractId]: Set[(Cid2, Identifier)] =
+    fold(Set.empty[(Cid2, Identifier)]) {
+      case (acc, (_, Node.Exercise(coid, tid, _, _, _, _, _, _, _, _, _, _, _, _, _))) =>
+        acc + ((coid, tid))
+      case (acc, (_, Node.Fetch(coid, tid, _, _, _, _, _, _))) =>
+        acc + ((coid, tid))
+      case (acc, (_, Node.LookupByKey(tid, _, Some(coid), _))) =>
+        acc + ((coid, tid))
       case (acc, _) => acc
-    } -- localContracts.keySet
+    }.filterNot(x => localContracts.keySet.contains(x._1))
```

Comment:
This is a public API (by KV at least). You cannot change it like that.

Reply:
Is there any other way I can get the template id? It's only used for the error message, but the message is fairly meaningless if you don't know the template of the inactive contract.

Reply:
Not really :(. i would probably copy/adapt your new `inputContracts` method directly you wrote in the Scenario Ledger.

 ```
 def tmplId(tx: Tx, coid: ContractId): Option[TypeConName] =
      transaction.nodes.values.collectFirst {
        case Node.Create(`coid`, tid, _, _, _, _, _, _) => tid
        case Node.Exercise(`coid`, tid, _, _, _, _, _, _, _, _, _, _, _, _, _) => tid
        case Node.Fetch(`coid`, tid, _, _, _, _, _, _) => tid
        case Node.LookupByKey(tid, _, Some(`coid`), _) => tid
      }
```

Reply:
@carlpulley-da could you please take care of this?